engine-hashchain/src/merkle.rs

```diff
@@ -0,0 +1,367 @@
+use aurora_engine_sdk::keccak;
+use aurora_engine_types::{
+    borsh::{self, BorshDeserialize, BorshSerialize},
+    types::RawH256,
+    Vec,
+};
+
+/// Stream Compact Merkle Tree.
+/// It can be feed by a stream of hashes (leaves) adding them to the right of the tree.
+/// Internally, compacts full binary subtrees maintaining only the growing branch.
+/// Space used is O(log n) where n is the number of leaf hashes added. It is usually less.
+#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq)]
+pub struct StreamCompactMerkleTree {
+    /// Complete binary merkle subtrees.
+    /// Left subtrees are strictly higher (bigger).
+    /// Subtrees are compacted right to left when two consecutives have same height.
+    subtrees: Vec<CompactMerkleSubtree>,
+}
+
+impl StreamCompactMerkleTree {
+    pub const fn new() -> Self {
+        Self {
+            subtrees: Vec::new(),
+        }
+    }
+
+    /// Returns `true` if no data has been added to the tree.
+    pub fn is_empty(&self) -> bool {
+        self.subtrees.is_empty()
+    }
+
+    /// Adds a leaf hash to the right of the tree.
+    /// For n leaves hashes added, a single call to this function is O(log n),
+    /// but the amortized time per each of the n calls is O(1).
+    pub fn add(&mut self, leaf_hash: RawH256) {
+        let leaf_subtree = CompactMerkleSubtree {
+            height: 1,
+            hash: leaf_hash,
+        };
+        self.subtrees.push(leaf_subtree);
+
+        // compact subtrees from right to left
+        let mut index = self.subtrees.len() - 1;
+
+        while index >= 1 {
+            debug_assert_eq!(index, self.subtrees.len() - 1);
+
+            let right_subtree = &self.subtrees[index];
+            let left_subtree = &self.subtrees[index - 1];
+
+            // same height means they are siblings so we can compact them
+            if left_subtree.height == right_subtree.height {
+                let father_subtree = CompactMerkleSubtree {
+                    height: left_subtree.height + 1,
+                    hash: keccak(&[left_subtree.hash, right_subtree.hash].concat()).0,
+                };
+
+                self.subtrees.pop();
+                self.subtrees.pop();
+                self.subtrees.push(father_subtree);
```

Comment:
I guess we could get rid of one `pop` and `push` here by replacing:
```rust
self.subtrees.pop();
self.subtrees.push(father_subtree);
```
to
```rust
*(self.subtrees.last_mut().unwrap()) = father_subtree;
```
`unwrap` is unsafe here because of `while index >= 1`.

Reply:
Done in https://github.com/aurora-is-near/aurora-engine/pull/816/commits/8e864578831aed9e8278e596689f926db751e035